### PR TITLE
planner: fix wrong aggregate pruning for some cases

### LINKED ISSUE #25202: dml: wrong result in subquery with `order by`  clause
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
create table a(b float);
SELECT avg(2) FROM(SELECT min(c) FROM a JOIN(SELECT 1 c) d ORDER BY b) e;
```

### 2. What did you expect to see? (Required)
```
+--------+
| avg(2) |
+--------+
| 2.0000 |
+--------+
1 row in set (0.00 sec)
```
### 3. What did you see instead (Required)
```
+--------+
| avg(2) |
+--------+
|   NULL |
+--------+
1 row in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)
```
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                            |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.1.0-alpha-167-ga8ad9b7d3
Edition: Community
Git Commit Hash: a8ad9b7d3aa53d0181422d2b22fa8d57b7d56db1
Git Branch: master
UTC Build Time: 2021-06-05 07:04:28
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by lzmhhh123**
Signed-off-by: lzmhhh123 <lzmhhh123@gmail.com>

### What problem does this PR solve?

Issue Number: close #25202

### What is changed and how it works?

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

- fix wrong aggregate pruning for some cases

### CODE REVIEW

File: planner/core/rule_column_pruning.go [hunk 1]
```diff
@@ -106,7 +109,7 @@ func (la *LogicalAggregation) PruneColumns(parentUsedCols []*expression.Column)
 		aggrFunc.OrderByItems, cols = pruneByItems(aggrFunc.OrderByItems)
 		selfUsedCols = append(selfUsedCols, cols...)
 	}
-	if len(la.AggFuncs) == 0 {
+	if len(la.AggFuncs) == 0 || (!allFirstRow && allRemainFirstRow) {
```
Comment (winoros) [inline planner/core/rule_column_pruning.go:112]: Add comments why we need to add this check

Comment (winoros) [inline planner/core/rule_column_pruning.go:112]: And when the `len(la.AggFuncs)` is empty, why the `allRemainFirstRow` can be true?

Comment (lzmhhh123) [reply planner/core/rule_column_pruning.go:112]: Comments are below this line. 
https://github.com/pingcap/tidb/pull/25289/files#diff-ae586e551e5a201a40c83192b7dc23cc2778739ef522fcb02e4fb7fac2125c99R113

Comment (rebelice) [reply planner/core/rule_column_pruning.go:112]: These two conditions are not AND, but OR.

**Review by rebelice — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**